src/rust/engine/fs/src/store.rs

```diff
@@ -256,54 +279,201 @@ mod local {
+    /// (excluding lmdb overhead).
+    ///
+    /// Returns the size it was shrunk to, which may be larger than target_bytes.
+    ///
+    pub fn shrink(&self, target_bytes: usize) -> Result<usize, String> {
+      let mut used_bytes: usize = 0;
+      let mut fingerprints_by_expired_ago = BinaryHeap::new();
+
+      self
+        .inner
+        .env
+        .begin_rw_txn()
+        .and_then(|mut txn| {
+          {
+            self.aged_fingerprints(
+              &txn,
+              self.inner.file_database,
+              &mut used_bytes,
+              &mut fingerprints_by_expired_ago,
+            );
+            self.aged_fingerprints(
+              &txn,
+              self.inner.directory_database,
+              &mut used_bytes,
+              &mut fingerprints_by_expired_ago,
+            );
+          }
+          while used_bytes > target_bytes {
+            let aged_fingerprint = fingerprints_by_expired_ago.pop().expect(
+              "lmdb corruption detected, sum of size of blobs exceeded stored blobs",
+            );
+            if aged_fingerprint.expired_seconds_ago == 0 {
+              // Ran out of expired blobs - everything remaining is leased and cannot be collected.
+              return Err(MapFull);
+            }
+            txn
+              .del(
+                aged_fingerprint.database,
+                &aged_fingerprint.fingerprint.as_ref(),
+                None,
+              )
+              .expect("Failed to delete lmdb blob");
+            used_bytes -= aged_fingerprint.size_bytes;
+          }
+          txn.commit()
+        })
+        .or_else(|e| match e {
+          MapFull => Ok(()),
+          e => Err(format!("Error shrinking store: {:?}", e)),
+        })?;
+      Ok(used_bytes)
+    }
+
+    fn aged_fingerprints<T>(
+      &self,
+      txn: &T,
+      database: Database,
+      used_bytes: &mut usize,
+      fingerprints_by_expired_ago: &mut BinaryHeap<AgedFingerprint>,
+    ) where
+      T: Transaction,
+    {
+      let mut cursor = txn.open_ro_cursor(database).expect(
+        "Failed to open lmdb read cursor",
+      );
+      for (key, bytes) in cursor.iter() {
+        *used_bytes = *used_bytes + bytes.len();
+
+        // Random access into the lease_database is slower than iterating, but hopefully garbage
+        // collection is rare enough that we can get away with this, rather than do two passes here
+        // (either to populate leases into pre-populated AgedFingerprints, or to read sizes when
+        // we delete from lmdb to track how much we've freed).
+        let lease_until_unix_timestamp = txn
+          .get(self.inner.lease_database, &key)
+          .map(|b| LittleEndian::read_u64(b))
+          .unwrap_or_else(|e| match e {
+            NotFound => 0,
+            e => panic!("Error reading lease, probable lmdb corruption: {:?}", e),
+          });
+
+        let leased_until = time::UNIX_EPOCH + time::Duration::from_secs(lease_until_unix_timestamp);
```

Comment:
To avoid needing hungarian notation for the variable, could execute the entire conversion into a `SystemTime` in the `map(|b| LittleEndian::read_u64(b))`. But perhaps that just moves the problem.

Reply:
Yeah, I'm not convinced it can go away entirely...